### PR TITLE
Sockets - Shubha

### PR DESCRIPTION
**Pull request by shubha-rajan**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | The most significant design challenge was avoiding tightly coupled classes. When I first started writing my code, I directly referenced instantiation methods for other classes in a lot of my classes. I quickly noticed that making changes to instantiate method signatures lead to an absurd amount of errors in my code, so I decided to use some of the strategies in this week's reading to attempt to decouple my code. I saw the immediate impact of this when making changes lead to only 8-10 errors or failures as opposed to 30-50.
What was a design decision you made that changed over time over the project? | Originally, I was storing collections of reservations and blocks in an array, but I made the decision to use hashes instead since I wanted to simplify reservation lookup. Originally I had the idea to use names or group names as keys for the hashes, but I ran into the issue of two reservations with the same name, where one would overlap with another. I then decided to generate unique IDs for reservation lookup instead and wrote a method which used the faker gem to generate random alphanumeric strings.
What was a concept you gained clarity on, or a learning that you'd like to share? | I'm still not super confident in the idea of loosely vs tightly coupled code, but I did gain clarity (mostly through the frustration of tiny changes resulting in 47 errors when running rake) in how avoiding things like chaining methods of other classes, or directly indexing into complex data structures can result in simpler and more flexible code. I have a tendency to chain a lot of methods when writing code, so this assignment was a good exercise in writing methods that isolate when once class calls a method in another class, and seeing how methods in different classes "talk" to eachother. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? |  An example of a nominal test would be checking whether a reservation can be made when a room has a reservation where the dates do not overlap at all. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | An example of an edge case test that I wrote for this assignment is a test to check whether an exception is raised when I try to reserve a room from a block when it is not part of the block.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I'm still getting used to the TDD workflow, but for this assignment, I found writing tests before code really helpful. In many cases, the user stories either directly or indirectly translated into tests that I could write, and the tests helped clarify how exactly I want my classes to behave. I didn't really write pseudocode, but I did roughly map out what classes I wanted to have, what information each class had access to, and what information each class could send to other classes. I drew this out in a UML diagram, which I used as a roadmap in writing my code, and I started with the class that directly interacted with other classes the least (DateRange) and worked my way up from there.

### CODE REVIEW

File: lib/block.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+require_relative "err/errors"
+
+module HotelSystem
+  class Block
+    attr_reader :rooms, :date_range, :discount_rate, :reservations, :id
+
+    def initialize(rooms:, date_range:, discount_rate:, id:)
+      if rooms.length > 5 || rooms.length < 1
+        raise BlockError, "A block must contain 1 to 5 rooms"
```
Comment (dHelmgren) [inline lib/block.rb]: Does a block of 1 make sense?

File: lib/err/errors.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class DateRangeError < StandardError
+end
+
+class BlockError < StandardError
```
Comment (dHelmgren) [inline lib/err/errors.rb:4]: Cool!

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      room.rate = new_rate
+    end
+
+    private
+
+    # Private helper methods:
+
+    def generate_id
+      id = Faker::Alphanumeric.unique.alphanumeric 10
+      return id.upcase
```
Comment (dHelmgren) [inline lib/hotel.rb]: This might ruin the uniqueness, make sure you check!

File: refactors.txt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+1. ADD MORE TESTS. ALWAYS MORE TESTS.
```
Comment (dHelmgren) [inline refactors.txt:1]: lmao!

**Comment by dHelmgren** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | Yes
Classes are loosely coupled | Yes
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | Good
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | Good
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | good
**Fundamentals** |
Names variables, classes and modules appropriately | yes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace |yes 
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | My only comment is this: it's okay to write beefier functions! A lot of the methods you wrote for this are tiny and digestible, but it means that someone reading the code has to jump around a lot at times. Food for thought!
